File: cpp/trex_selector_methods/trex_screening/demo_trex_scr_06_mc_sim_solvers/demo_trex_scr_06_mc_sim_solvers.cpp
// ==============================================================================
// demo_trex_scr_06_mc_sim_solvers.cpp
// ==============================================================================
/**
 * @file demo_trex_scr_06_mc_sim_solvers.cpp
 *
 * @brief Screen-TRex Selector — solver comparison: TLARS, TAFS (ρ=0.3), TOMP.
 *
 * @details
 *  Compares the three T-Rex solver backends across both Screen-TRex method
 *  variants (ordinary Phi > 0.5 threshold and bootstrap-CI).
 *
 *  Demo 01 — Monte Carlo SNR sweep: three solvers, ordinary Screen-TRex only.
 *             Parallelised via OpenMP.
 *  Demo 02 — Monte Carlo SNR sweep: three solvers × ordinary + bootstrap (6 series).
 *             Parallelised via OpenMP.
 *
 *  In-memory variant (X and D matrices held in RAM).
 */
// ==============================================================================

// std includes
#include <iostream>
#include <map>
#include <random>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// TRex includes
#include <utils/datageneration/utils_datagen.hpp>

// Screen-TRex demo simulation utilities
#include "trex_screening_simulation_utils.hpp"

// ==============================================================================
// Namespace aliases
// ==============================================================================

namespace datagen = trex::utils::datageneration::datagen;
using namespace scr_demo;

// ==============================================================================
// Local helpers — combined solver × method-variant descriptors
// ==============================================================================

namespace {

/**
 * @brief Combined descriptor for a Screen-TRex solver × method variant.
 *
 * @details Extends the library's ScreenMethodInfo with the solver backend
 *          and the AFS penalty parameter so that the outer MC loop can build
 *          both the TRexControlParameter and the ScreenTRexControlParameter
 *          from a single object.
 */
struct SolverVariant {
    std::string       name;                             ///< Display name
    SolverTypeForTRex solver_type;                      ///< T-Rex solver backend
    double            rho_afs   = 0.0;                  ///< AFS penalty (TAFS only)
    ScreenTRexMethod  method    = ScreenTRexMethod::TREX; ///< Screen-TRex variant
    bool              bootstrap = false;                 ///< Use bootstrap-CI threshold
};

/** @brief Three-solver list — ordinary Screen-TRex only. */
inline std::vector<SolverVariant> ordinary_solver_variants()
{
    return {
        {"TLARS (Ordinary)",    SolverTypeForTRex::TLARS, 0.0,
         ScreenTRexMethod::TREX, false},
        {"TAFS-0.3 (Ordinary)", SolverTypeForTRex::TAFS,  0.3,
         ScreenTRexMethod::TREX, false},
        {"TOMP (Ordinary)",     SolverTypeForTRex::TOMP,  0.0,
         ScreenTRexMethod::TREX, false},
    };
}

/** @brief Six combined variants — three solvers × ordinary + bootstrap. */
inline std::vector<SolverVariant> all_solver_variants()
{
    std::vector<SolverVariant> v;
    v.reserve(6);
    for (const bool bs : {false, true}) {
        const std::string suffix = bs ? " (Bootstrap)" : " (Ordinary)";
        v.push_back({"TLARS"    + suffix, SolverTypeForTRex::TLARS, 0.0,
                     ScreenTRexMethod::TREX, bs});
        v.push_back({"TAFS-0.3" + suffix, SolverTypeForTRex::TAFS,  0.3,
                     ScreenTRexMethod::TREX, bs});
        v.push_back({"TOMP"     + suffix, SolverTypeForTRex::TOMP,  0.0,
                     ScreenTRexMethod::TREX, bs});
    }
    return v;
}

/**
 * @brief Build a TRexControlParameter configured for the given SolverVariant.
 *
 * @param sv  Solver variant whose type and rho_afs are applied to the default
 *            TRexControlParameter produced by make_trex_control().
 * @return    Ready-to-use TRexControlParameter.
 */
inline TRexControlParameter make_trex_ctrl_for(const SolverVariant& sv)
{
    auto ctrl = make_trex_control();
    ctrl.solver_type           = sv.solver_type;
    ctrl.solver_params.rho_afs = sv.rho_afs;
    ctrl.tloop_stagnation_stop = true;
    ctrl.tloop_max_stagnant_steps = 5;
    return ctrl;
}

/**
 * @brief Convert a SolverVariant list to ScreenMethodInfo (name only).
 *
 * @details Used to pass the ordered method list to save_and_print_mc_results(),
 *          which needs names to index into the FDR/TPR maps.
 */
inline std::vector<ScreenMethodInfo>
to_screen_method_info(const std::vector<SolverVariant>& variants)
{
    std::vector<ScreenMethodInfo> info;
    info.reserve(variants.size());
    for (const auto& sv : variants)
        info.push_back({sv.name, sv.method, sv.bootstrap});
    return info;
}

} // anonymous namespace


// ==============================================================================
// Demo 01: MC SNR sweep — three solvers, ordinary Screen-TRex
// ==============================================================================

/**
 * @brief SNR-sweep Monte Carlo comparing TLARS, TAFS (ρ=0.3), and TOMP.
 *
 * @details Uses ordinary Screen-TRex (Phi > 0.5 threshold) for all three
 *          solvers.  Reports FDR, TPR, and estimated FDR averaged over num_MC
 *          Monte Carlo runs at each SNR grid point.  Parallelised via OpenMP.
 *
 * @param num_MC   Number of Monte Carlo runs per (solver, SNR) pair.
 * @param high_dim If true, use high-dimensional setting (n=300, p=1000).
 * @param rnd_coef If true, draw support coefficients from N(0,1) each run.
 */
void demo_SolverComparison_MC(std::size_t num_MC, bool high_dim, bool rnd_coef)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Demo: Screen-TRex MC — Solver Comparison (TLARS / TAFS-0.3 / TOMP)");

    const std::size_t n = high_dim ? 300  : 1000;
    const std::size_t p = high_dim ? 1000 :  300;
    std::cout << (high_dim ? "High-dimensional (p > n)"
                           : "Low-dimensional (n > p)") << "\n";

    const std::size_t         support_size = 10;
    const std::vector<double> snr_values   =
        {0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0};

    const auto variants = ordinary_solver_variants();
    const auto methods  = to_screen_method_info(variants);

    // Storage: variant name → SNR-indexed accumulator
    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& sv : variants) {
        const auto sz        = static_cast<Eigen::Index>(snr_values.size());
        est_fdr_map[sv.name] = Eigen::VectorXd::Zero(sz);
        fdr_map    [sv.name] = Eigen::VectorXd::Zero(sz);
        tpr_map    [sv.name] = Eigen::VectorXd::Zero(sz);
    }

    // -----------------------------------------------------------------------
    // Solver loop
    // -----------------------------------------------------------------------
    for (const auto& sv : variants) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Solver: " << sv.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto trex_ctrl   = make_trex_ctrl_for(sv);
        auto screen_ctrl = make_screen_control(sv.method, sv.bootstrap);

        // SNR loop -----------------------------------------------------------
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {
            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_iid_dgp(n, p, support_size, snr, rnd_coef, seed);
            };

            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;

            std::ostringstream lbl;
            lbl << sv.name << "  SNR=" << std::fixed << std::setprecision(2) << snr;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto ei = static_cast<Eigen::Index>(snr_idx);
            est_fdr_map[sv.name](ei) = result.avg_est_fdr;
            fdr_map    [sv.name](ei) = result.avg_fdr;
            tpr_map    [sv.name](ei) = result.avg_tpr;
        }
    }

    const std::string file_stem =
        "scr_solvers_snr_n" + std::to_string(n)
        + "_p" + std::to_string(p)
        + "_s" + std::to_string(support_size);
    save_and_print_mc_results(num_MC, file_stem, snr_values, methods,
                              fdr_map, tpr_map, est_fdr_map);
}


// ==============================================================================
// Demo 02: MC SNR sweep — three solvers × ordinary + bootstrap
// ==============================================================================

/**
 * @brief SNR-sweep Monte Carlo over all six (solver × method-variant) combinations.
 *
 * @details Extends Demo 02 by also running bootstrap-CI Screen-TRex for each
 *          solver, producing six series:
 *            TLARS / TAFS-0.3 / TOMP  ×  Ordinary / Bootstrap.
 *          Parallelised via OpenMP.
 *
 * @param num_MC   Number of Monte Carlo runs per (variant, SNR) pair.
 * @param high_dim If true, use high-dimensional setting (n=300, p=1000).
 * @param rnd_coef If true, draw support coefficients from N(0,1) each run.
 */
void demo_SolverAndMethod_MC(std::size_t num_MC, bool high_dim, bool rnd_coef)
{
    std::cout.setf(std::ios::unitbuf);
    cdiag::print_section_header(
        "Demo: Screen-TRex MC — Solver × Method Comparison (6 series)");

    const std::size_t n = high_dim ? 300  : 1000;
    const std::size_t p = high_dim ? 1000 :  300;
    std::cout << (high_dim ? "High-dimensional (p > n)"
                           : "Low-dimensional (n > p)") << "\n";

    const std::size_t         support_size = 10;
    const std::vector<double> snr_values   =
        {0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0};

    const auto variants = all_solver_variants();
    const auto methods  = to_screen_method_info(variants);

    // Storage: variant name → SNR-indexed accumulator
    std::map<std::string, Eigen::VectorXd> est_fdr_map, fdr_map, tpr_map;
    for (const auto& sv : variants) {
        const auto sz        = static_cast<Eigen::Index>(snr_values.size());
        est_fdr_map[sv.name] = Eigen::VectorXd::Zero(sz);
        fdr_map    [sv.name] = Eigen::VectorXd::Zero(sz);
        tpr_map    [sv.name] = Eigen::VectorXd::Zero(sz);
    }

    // -----------------------------------------------------------------------
    // Variant loop (solver × method-variant)
    // -----------------------------------------------------------------------
    for (const auto& sv : variants) {
        std::cout << std::string(70, '=') << "\n";
        std::cout << "Variant: " << sv.name << "\n";
        std::cout << std::string(70, '=') << "\n";

        auto trex_ctrl   = make_trex_ctrl_for(sv);
        auto screen_ctrl = make_screen_control(sv.method, sv.bootstrap);

        // SNR loop -----------------------------------------------------------
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {
            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> ScrDGPData {
                return make_iid_dgp(n, p, support_size, snr, rnd_coef, seed);
            };

            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;

            std::ostringstream lbl;
            lbl << sv.name << "  SNR=" << std::fixed << std::setprecision(2) << snr;
            const auto result = run_mc_trials_screen_trex(
                num_MC, lbl.str(), make_data, trex_ctrl, screen_ctrl, base_seed);

            const auto ei = static_cast<Eigen::Index>(snr_idx);
            est_fdr_map[sv.name](ei) = result.avg_est_fdr;
            fdr_map    [sv.name](ei) = result.avg_fdr;
            tpr_map    [sv.name](ei) = result.avg_tpr;
        }
    }

    const std::string file_stem =
        "scr_solver_method_snr_n" + std::to_string(n)
        + "_p" + std::to_string(p)
        + "_s" + std::to_string(support_size);
    save_and_print_mc_results(num_MC, file_stem, snr_values, methods,
                              fdr_map, tpr_map, est_fdr_map);
}


// ==============================================================================
// main
// ==============================================================================

int main()
{
    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);

    // -------------------------------------------------------------------------
    // Demo 01: MC SNR sweep — three solvers, ordinary Screen-TRex
    // -------------------------------------------------------------------------
    demo_SolverComparison_MC(200, true, false);

    // -------------------------------------------------------------------------
    // Demo 02: MC SNR sweep — three solvers × ordinary + bootstrap (6 series)
    // -------------------------------------------------------------------------
    demo_SolverAndMethod_MC(200, true, false);

    return 0;
}

File: Python/trex_selector_methods/trex_screening/demo_trex_scr_06_mc_sim_solvers/demo_trex_scr_06_mc_sim_solvers.py
# ==============================================================================
# demo_trex_scr_06_mc_sim_solvers.py
# ==============================================================================
#
# Screen-TRex Selector, demo 06: solver-backend comparison (TLARS / TAFS / TOMP).
#
# Screening thresholds the dummy-based voting proportion Phi_j, but Phi itself is
# produced by an underlying T-Rex solver. This demo asks whether the choice of
# solver backend changes screening performance:
#   * TLARS    - terminating LARS (the default),
#   * TAFS-0.3 - terminating adaptive forward selection (rho_afs = 0.3),
#   * TOMP     - terminating orthogonal matching pursuit,
# each combined with Ordinary and Bootstrap-CI screening. The same i.i.d.
# Gaussian design as demo 01 is used, so the numbers are directly comparable.
#
# Mirrors:
#   cpp/trex_selector_methods/trex_screening/demo_trex_scr_06_mc_sim_solvers/
#       demo_trex_scr_06_mc_sim_solvers.cpp
#
# Demo content (two Monte Carlo SNR sweeps, matching the cpp Demos 01-02):
#   Part 1: three solvers, Ordinary screening only (3 series).
#   Part 2: three solvers x {Ordinary, Bootstrap-CI} (6 series).
#
# Downscaled from cpp: the cpp demo uses num_MC = 200; committed default is
# smaller here for a practical Python runtime (override with SCR_NUM_MC). Support
# indices are 0-based.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from trex_scr_common import run_mc_screen, save_and_print_scr_mc

_OUT_DIR = os.path.join(_THIS_DIR, "simulation_results", "data")

# Committed defaults (cpp uses num_MC = 200; downscaled here for Python runtime).
_NUM_MC      = int(os.environ.get("SCR_NUM_MC", 50))
_NUM_WORKERS = int(os.environ.get("SCR_NUM_WORKERS", 6))

N, P, SUPPORT_SIZE = 300, 1000, 10
SNR_GRID = [0.01, 0.1, 0.2, 0.5, 0.6, 1.0, 2.0, 5.0]


def _sv(name, solver, rho_afs=None, bootstrap=False):
    """A solver x screening-variant descriptor (mirrors cpp SolverVariant)."""
    return {"name": name, "solver": solver, "rho_afs": rho_afs, "bootstrap": bootstrap}


def _ordinary_solver_variants():
    """Three solvers, Ordinary screening only (mirrors ordinary_solver_variants())."""
    return [
        _sv("Screen-TRex Ordinary: TLARS",    "TLARS"),
        _sv("Screen-TRex Ordinary: TAFS-0.3", "TAFS", rho_afs=0.3),
        _sv("Screen-TRex Ordinary: TOMP",     "TOMP"),
    ]


def _all_solver_variants():
    """Six combined variants - three solvers x {Ordinary, Bootstrap-CI}."""
    variants = []
    for bs in (False, True):
        prefix = "Screen-TRex Bootstrap-CI: " if bs else "Screen-TRex Ordinary: "
        variants.append(_sv(prefix + "TLARS",    "TLARS",           bootstrap=bs))
        variants.append(_sv(prefix + "TAFS-0.3", "TAFS", rho_afs=0.3, bootstrap=bs))
        variants.append(_sv(prefix + "TOMP",     "TOMP",            bootstrap=bs))
    return variants


def run_solver_sweep(title, variants, file_stem, num_MC=_NUM_MC, num_workers=_NUM_WORKERS):
    """MC sweep over a list of solver x method variants (SNR grid)."""
    print("\n" + "=" * 70)
    print(title)
    print("=" * 70)
    print("High-dimensional (p > n)")
    print(f"n = {N}, p = {P}, s = {SUPPORT_SIZE}, num_MC = {num_MC}\n")

    method_names = [v["name"] for v in variants]

    def init():
        return {nm: [0.0] * len(SNR_GRID) for nm in method_names}
    fdr_map, tpr_map, est_fdr_map = init(), init(), init()

    for v in variants:
        print("=" * 70)
        print(f"Variant: {v['name']}")
        print("=" * 70)
        # make_trex_ctrl_for(): stagnation stop enabled for every backend.
        trex_args = {"solver": v["solver"], "use_memory_mapping": False, "K": 20,
                     "tloop_stagnation_stop": True, "tloop_max_stagnant_steps": 5,
                     "rho_afs": v["rho_afs"]}
        screen_args = {"trex_method": "TREX", "bootstrap": v["bootstrap"]}
        for si, snr in enumerate(SNR_GRID):
            dgp_args = {"dgp_kind": "iid", "n": N, "p": P,
                        "support_size": SUPPORT_SIZE, "snr": snr, "rnd_coef": False}
            base_seed = 24 + si * 1000        # mirrors cpp base_seed = 24 + snr_idx * 1000
            res = run_mc_screen(num_MC, f"{v['name']}  SNR={snr:.2f}",
                                dgp_args, trex_args, screen_args,
                                base_seed=base_seed, max_workers=num_workers)
            fdr_map[v["name"]][si] = res["avg_fdr"]
            tpr_map[v["name"]][si] = res["avg_tpr"]
            est_fdr_map[v["name"]][si] = res["avg_est_fdr"]

    save_and_print_scr_mc(num_MC, _OUT_DIR, file_stem, SNR_GRID,
                          method_names, fdr_map, tpr_map, est_fdr_map)


# ==============================================================================
# Main
# ==============================================================================

def main():
    # Part 1: three solvers, ordinary Screen-TRex
    run_solver_sweep(
        "Demo: Screen-TRex MC - Solver Comparison (TLARS / TAFS-0.3 / TOMP)",
        _ordinary_solver_variants(),
        f"scr_solvers_snr_n{N}_p{P}_s{SUPPORT_SIZE}")

    # Part 2: three solvers x {Ordinary, Bootstrap-CI} (6 series)
    run_solver_sweep(
        "Demo: Screen-TRex MC - Solver x Method Comparison (6 series)",
        _all_solver_variants(),
        f"scr_solver_method_snr_n{N}_p{P}_s{SUPPORT_SIZE}")

    print("\ndemo_trex_scr_06_mc_sim_solvers.py complete.")


if __name__ == "__main__":
    main()
